source/ddbus/router.d

```diff
@@ -105,22 +105,32 @@ class MessageRouter {
 
   void setHandler(Ret, Args...)(MessagePattern patt, Ret delegate(Args) handler) {
     void handlerWrapper(Message call, Connection conn) {
+      import ddbus.exception : DBusErrorReturn;
       Tuple!Args args = call.readTuple!(Tuple!Args)();
-      auto retMsg = call.createReturn();
-
-      static if (!is(Ret == void)) {
-        Ret ret = handler(args.expand);
-        static if (is(Ret == Tuple!T, T...)) {
-          retMsg.build!T(ret.expand);
-        } else {
-          retMsg.build(ret);
-        }
-      } else {
-        handler(args.expand);
-      }
 
-      if (!patt.signal) {
-        conn.send(retMsg);
+      try {
+	auto retMsg = call.createReturn();
+
+	static if (!is(Ret == void)) {
+	  Ret ret = handler(args.expand);
+	  static if (is(Ret == Tuple!T, T...)) {
+	    retMsg.build!T(ret.expand);
+	  } else {
+	    retMsg.build(ret);
+	  }
+	} else {
+	  handler(args.expand);
+	}
+
+	if (!patt.signal) {
+	  conn.send(retMsg);
+	}
+      } catch(DBusErrorReturn e) {
+	auto errRet = Message(dbus_message_new_error(call.msg,
+	                      e.errorName.toStringz, e.msg.toStringz));
+	if (!patt.signal) {
+	  conn.send(errRet);
+	}
```

Comment:
these 3 lines (and the ones ending in the catch expression) can be moved out of the try-catch to avoid code duplication